ci(jest): Read jest-balance.json from github artifacts dynamically during frontend CI Jest tests

ryan953 opened the pull request:
Previously we had `jest-balance.json` saved into the codebase. It was managed/updated by a github action which would generate the file, then open a PR to merge in the json. 
This was annoying because someone had to manually merge the updates, which usually didn't happen quickly. Also the json was not formatted (until https://github.com/getsentry/sentry/pull/100678) so it was impossible to read the PR. It's still impossible to read because the number change slightly after each run, so the diff is always huge. 

So basically we were blindly merging this JSON into the repo ~weekly or ~monthly, whenever someone got around to it. The 'balance' system works well through, we currently use 4 runners and they complete within within a few seconds of each other. (see image at bottom)

As of https://github.com/getsentry/sentry/pull/100683 that action is uploading the `jest-balance.json` as an artifact, in addition to opening a PR...
With this change we'll pull the artifact down in CI, and not bother to read it from the codebase anymore. This means that the 'create PR' step can go away! No more manual steps, and we'll always have balanced CI runs.

**How well does Jest Balance work for us?**
it works pretty well! you can see from this chart that p75 times are very stable for the jest workers, with the exception of between july 13 to 25th. During that time the balancer had an outage and we split tests up naively, this caused CI runtime to suffer while we waited for the slowest shard.
you can also see how https://github.com/getsentry/sentry/pull/97755 brought the duration of eslint down, so jest is firmly the most expensive step.
<img width="729" height="276" alt="SCR-20251001-ogwj" src="https://github.com/user-attachments/assets/088c463d-c432-4cee-8053-958f8624f988" />

ryan953 committed:
ci(jest): Read jest-balance.json from github artifacts dynamically as neede

ryan953 committed:
Remove create-PR step

ryan953 committed:
try to find the artifact without specifying a run-id

ryan953 committed:
get the github token, so we can find artifacts from other runs

ryan953 committed:
fix path

ryan953 committed:
use a different action to download

ryan953 committed:
only use dawidd6/action-download-artifact

ryan953 committed:
search wider

ryan953 committed:
try the name of the workflow, not filename

ryan953 committed:
searching for artifacts

ryan953 committed:
back to workflow file name

ryan953 committed:
only master branch

ryan953 committed:
dont specify workflow, we will find it

ryan953 committed:
hard-code the workflow id for perf

ryan953 committed:
tidy up

scttcper commented on .github/workflows/frontend.yml:
```diff
@@ -190,6 +190,26 @@ jobs:
+
+      - name: Download jest-balance.json
+        id: download-artifact
+        uses: dawidd6/action-download-artifact@v11
```
i think our security team wants us to use commits instead of tag resolution

ryan953 committed:
use sha to reference dawidd6/action-download-artifact

ryan953 committed:
delete the json so jest.config.js falls through properly to a different sharding strategy

if we have an empty file the the tests with "Speculated suite duration is <= 0" and not run at all

ryan953 committed:
rm extra step

ryan953 committed:
rm console.log
